feat(auth): cancel subscriptions to plan script

julianpoy opened the pull request:
## Because

* We need a way to bulk cancel subscriptions to a given plan

## This pull request

* Creates a script that can cancel subscriptions to a given plan with a few configuration options

## Issue that this pull request solves

Closes FXA-7124

IvoJP commented on packages/fxa-auth-server/scripts/cancel-subscriptions-to-plan/cancel-subscriptions-to-plan.ts:
```diff
@@ -0,0 +1,261 @@
+        : subscription.latest_invoice.id;
+
+    const invoice = await this.enqueueRequest(() =>
+      this.stripe.invoices.retrieve(latestInvoiceId)
```
Is this call to retrieve an invoice necessary if `latest_invoice` is not a string (i.e. its an invoice)?
Not sure if saving an extra stripe call on this script matters, but pointing this out in case it does.

julianpoy replied:
It's not technically necessary in the case that latest_invoice is not a string.

I do this because this object is from Firestore so I'm using a fair amount of caution here and re-fetching in case the last invoice in Firestore is stale, especially right before we issue a credit directly to the user's card.

IvoJP commented on packages/fxa-auth-server/scripts/cancel-subscriptions-to-plan/cancel-subscriptions-to-plan.ts:
```diff
@@ -0,0 +1,261 @@
+
+      console.log(subscriptionId);
+    } catch (e) {
+      console.error(subscriptionId, e);
```
If we are capturing the success to the report file, do we want to capture the errors to anything other than std err?

julianpoy replied:
Nope - since the subscription is likely unaffected, we mustn't include them in the CSV for marketing. I use stdout/stderr for all message parsing and analytics for the results of the script run. The stdout/stderr is captured by SRE, so if there are errors we'll be able to use the information dumped here. The cancellation of any particular subscription also shouldn't block the script from continuing.

IvoJP reviewed (approved):
Would have liked to see some return types on functions and maybe some `{}` even for single line `if` blocks but thats also the c#/java dev in me :( and we also don't have explicit rules requiring such so this is non-blocking.

Left a few questions that are also non-blocking so author's discretion on them.

LGTM.

julianpoy commented on packages/fxa-auth-server/scripts/cancel-subscriptions-to-plan/cancel-subscriptions-to-plan.ts:
```diff
@@ -0,0 +1,261 @@
+   */
+  async fetchSubsBatch(startAfter: string | null) {
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const subscriptionCollection = `${collectionPrefix}subscriptions`;
```
Not strictly necessary. In past scripts had been using other collections so it had more of a point. It could be removed, but also not hindering anything

julianpoy commented:
> Would have liked to see some return types on functions

Added! Definitely agree, we should have explicit return types on all functions. Adds readability, and improves TypeScript compiler performance! Win win.

> maybe some {} even for single line if blocks but thats also the c#/java dev in me :( and we also don't have explicit rules requiring such so this is non-blocking.

Completely understand the instinct here - I just don't agree on this one. Personally a fan of single line `if` blocks, and very much fits in with the JS/TS style imo. As always, I'm happy to have a discussion on this one, and if the team wants to put in a lint rule to block this style, I do not mind it (the linter will correct them for me hah!).

> Left a few questions that are also non-blocking so author's discretion on them.

I think I addressed everything, so going to merge to be ready for tag today, but happy to continue discussions on any points you feel necessary post-merge.
